lib/charms/layer/apache_zeppelin.py

```diff
@@ -144,18 +204,50 @@ def start(self):
             zeppelin_home = self.dist_config.path('zeppelin')
             # chdir here because things like zepp tutorial think ZEPPELIN_HOME
             # is wherever the daemon was started from.
-            os.chdir(zeppelin_home)
-            utils.run_as('ubuntu',
-                         '{}/bin/zeppelin-daemon.sh'.format(zeppelin_home),
-                         '--config', zeppelin_conf,
-                         'start')
+            with host.chdir(zeppelin_home):
+                utils.run_as('ubuntu',
+                             '{}/bin/zeppelin-daemon.sh'.format(zeppelin_home),
+                             '--config', zeppelin_conf,
+                             'start')
+            # wait up to 30s for server to start responding, lest API requests fail
+            self.wait_for_api(30)
+
+    def check_connect(self, addr, port):
+        try:
+            with socket.create_connection((addr, port), timeout=10):
+                return True
+        except OSError:
+            return False
+
+    def wait_for_api(self, timeout):
+        start = time.time()
+        while time.time() - start < timeout:
+            if self.check_connect('localhost', self.dist_config.port('zeppelin')):
+                return True
+            time.sleep(2)
+        raise utils.TimeoutError('Timed-out waiting for connection to Zeppelin')
+
+    def wait_for_stop(self, timeout):
+        start = datetime.now()
+        while utils.jps("zeppelin"):
+            time.sleep(1)
+            if datetime.now() - start > timeout:
```

Comment:
Another fix would just be to use time.time() for this. You're just measuring local system time, so you don't actually want the overhead of using datetime.